Sources/SwiftSyntaxMacroExpansion/MacroSystem.swift

```diff
@@ -666,6 +671,7 @@ private class MacroApplication<Context: MacroExpansionContext>: SyntaxRewriter {
   /// Store expanded extension while visiting member decls. This should be
   /// added to top-level 'CodeBlockItemList'.
   var extensions: [CodeBlockItemSyntax] = []
+  var expandedAttributes: [AttributeSyntax] = []
```

Comment:
> I think we might even do this based on the ID now and don’t need to resort to the position

I found that the id is still different. The `macroAttributes` method depends on the `attachedTo` node. In `visitAny` the `declSyntax`'s root is `SourceFileSyntax`, while in `visit`(for example [`visit(_ node: VariableDeclSyntax)`](https://github.com/apple/swift-syntax/blob/c3fab73792a5bbb70d38a5ea3ee9d0930915351a/Sources/SwiftSyntaxMacroExpansion/MacroSystem.swift#L919)) the node's root is `VariableDeclSyntax`(itself).

It seems that in [`visitChildren`](https://github.com/apple/swift-syntax/blob/c3fab73792a5bbb70d38a5ea3ee9d0930915351a/Sources/SwiftSyntax/generated/SyntaxRewriter.swift#L3889), the node will have a new layout and generate a new syntax without origin root info. Is this by design? This means if we use position to compare is also incorrect.

Reply:
I also delved into the implementation of `AttributeSyntax.position`. It's **the position of the start of this node's leading trivia**(not the root). For `Attribute`, it always represents its index in the `AttributeList`. So I don't think it's a proper way to compare two attributes by position

Reply:
I think I’ll need to play around with the code myself a little to figure out how we want to continue, but am too busy with other things at the moment. I’ll definitely come back to you but it might take a few weeks.

Reply:
Hi @ahoppen , are there any updates on this?

Reply:
I was super busy implementing background indexing in SourceKit-LSP over the [last few weeks](https://github.com/apple/sourcekit-lsp/pulls?q=is%3Apr+author%3Aahoppen+created%3A%3E2024-05-07+). I’m hoping that this will calm down now and I’ll allow me to get back to this PR this week.

Reply:
Sorry for the delay, I found some time now. Looks like we were really close, the only issue was that in `visit(_: VariableDeclSyntax)` we needed to call `macroAttributes(attachedTo:ofType:)` on the original node, not the one that we already processed using `super.visit`.

<details><summary>Diff</summary>
<p>

```diff
diff --git a/Sources/SwiftSyntaxMacroExpansion/MacroSystem.swift b/Sources/SwiftSyntaxMacroExpansion/MacroSystem.swift
index 23127070..f350c097 100644
--- a/Sources/SwiftSyntaxMacroExpansion/MacroSystem.swift
+++ b/Sources/SwiftSyntaxMacroExpansion/MacroSystem.swift
@@ -669,7 +669,7 @@ private class MacroApplication<Context: MacroExpansionContext>: SyntaxRewriter {
   /// Store expanded extension while visiting member decls. This should be
   /// added to top-level 'CodeBlockItemList'.
   var extensions: [CodeBlockItemSyntax] = []
-  var expandedAttributes: [AttributeSyntax] = []
+  /// Attributes that we are expecting to expand from `visitAny`. As macros are expanded, they should be removed from this list. Any attributes that are still left in this array after a `visitAny` call will generate diagnostics.
+  var attributesToExpand: [(attributeNode: AttributeSyntax, spec: MacroSpec)] = []
 
   init(
     macroSystem: MacroSystem,
@@ -707,6 +707,11 @@ private class MacroApplication<Context: MacroExpansionContext>: SyntaxRewriter {
       let attributedNode = node.asProtocol(WithAttributesSyntax.self),
       !attributedNode.attributes.isEmpty
     {
+      let previousAttributesToExpand = attributesToExpand
+      defer {
+        attributesToExpand = previousAttributesToExpand
+      }
+      attributesToExpand = self.macroAttributes(attachedTo: declSyntax)
       // Apply body and preamble macros.
       if let nodeWithBody = node.asProtocol(WithOptionalCodeBlockSyntax.self),
         let declNodeWithBody = nodeWithBody as? any DeclSyntaxProtocol & WithOptionalCodeBlockSyntax
@@ -720,14 +725,8 @@ private class MacroApplication<Context: MacroExpansionContext>: SyntaxRewriter {
       skipVisitAnyHandling.remove(Syntax(declSyntax))
 
       let attributesToRemove = self.macroAttributes(attachedTo: visitedNode)
-      for (attribute, spec) in attributesToRemove {
-        if let index = self.expandedAttributes.firstIndex(where: { expandedAttribute in
-          expandedAttribute.position == attribute.position
-        }) {
-          self.expandedAttributes.remove(at: index)
-        } else {
-          self.diagnosticForUnexpandedMacro(attribute: attribute, macroType: spec.type, attachedTo: declSyntax)
-        }
+      for attribute in attributesToExpand {
+        self.diagnosticForUnexpandedMacro(attribute: attribute.attributeNode, macroType: attribute.spec.type, attachedTo: declSyntax)
       }
       return AttributeRemover(removingWhere: { attributesToRemove.map(\.attributeNode).contains($0) }).rewrite(
         visitedNode
@@ -961,10 +960,10 @@ private class MacroApplication<Context: MacroExpansionContext>: SyntaxRewriter {
     )
   }
 
-  override func visit(_ node: VariableDeclSyntax) -> DeclSyntax {
-    var node = super.visit(node).cast(VariableDeclSyntax.self)
+  override func visit(_ originalNode: VariableDeclSyntax) -> DeclSyntax {
+    var node = super.visit(originalNode).cast(VariableDeclSyntax.self)
 
-    guard !macroAttributes(attachedTo: DeclSyntax(node), ofType: AccessorMacro.Type.self).isEmpty else {
+    guard !macroAttributes(attachedTo: DeclSyntax(originalNode), ofType: AccessorMacro.Type.self).isEmpty else {
       return DeclSyntax(node)
     }
 
@@ -1061,6 +1060,7 @@ extension MacroApplication {
     var result: [ExpandedNode] = []
 
     for macroAttribute in macroAttributes(attachedTo: decl, ofType: ofType) {
+      attributesToExpand = attributesToExpand.filter { $0.attributeNode != macroAttribute.attributeNode }
       do {
         if let expanded = try expandMacro(
           macroAttribute.attributeNode,
@@ -1068,7 +1068,6 @@ extension MacroApplication {
           macroAttribute.conformanceList
         ) {
           result += expanded
-          self.expandedAttributes.append(macroAttribute.attributeNode)
         }
       } catch {
         contextGenerator(Syntax(decl)).addDiagnostics(from: error, node: macroAttribute.attributeNode)
@@ -1216,7 +1215,7 @@ extension MacroApplication {
               from: newAccessors,
               indentationWidth: self.indentationWidth
             )
-            self.expandedAttributes.append(macro.attributeNode)
+            attributesToExpand = attributesToExpand.filter { $0.attributeNode != macro.attributeNode }
           }
         } else if let newAccessors = try expandAccessorMacroWithoutExistingAccessors(
           definition: macro.definition,
@@ -1239,7 +1238,7 @@ extension MacroApplication {
           } else {
             newAccessorsBlock = newAccessors
           }
-          self.expandedAttributes.append(macro.attributeNode)
+          attributesToExpand = attributesToExpand.filter { $0.attributeNode != macro.attributeNode }
         }
       } catch {
         contextGenerator(Syntax(storage)).addDiagnostics(from: error, node: macro.attributeNode)
```

</p>
</details>

Reply:
Looking at the code again, I don’t understand what I did myself 🤦🏽  Do you have a test at hand that shows the issue? And I’m getting the feeling that instead of working around things ad-hoc here, the right solution is to productize https://github.com/apple/swift-syntax/pull/2118, which will allow us to keep track of the node’s in the original tree.

Reply:
I've pushed my code. As we discussed above, unified diagnostic for unexpanded macros is already implemented. Because of the unexpected result of comparing two attributedNode, a lot of tests will fail.

Thinking back to the problem I encountered at the beginning, it was to compare whether two attributeNodes were the same in the origin syntax tree. So I agree with that #2118 should be the right solution.